examples/random/random_error.go

```diff
@@ -0,0 +1,55 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+
+	"github.com/codeship/go-retro"
+)
+
+const maxAttemps = 2
+
+var (
+	// ErrValueBelowThreshold is returned when the value is under the threshold.
+	// This error is retryable and will retry 5 times without sleeping
+	ErrValueBelowThreshold = retro.NewStaticRetryableError(errors.New("error: value is under the threshold"), maxAttemps, 0)
+)
+
+// Stores all ints from 0-99 under random unique keys in a map
+func main() {
+	for i := 0; i < 100; i++ {
+		var generatedValue int
+		err := retro.DoWithRetry(func() error {
```

Comment:
I think a more simplified way of looking at this would be to split the value and error handling functionality, and the debugging

```
err := retro.DoWithError(func() (err error) {
  defer func(){
      if err != nil {
	msg = fmt.Sprintf("%s Value %d is below the given threshold", msg, value)
	if i == maxAttemps {
	  msg = fmt.Sprintf("%s and reached maxAttemps", msg)
	}
      } else {
	msg = fmt.Sprintf("%s Value %d above the threshold - OK", msg, value)
     }
  }()
  generatedValue, err = genValue()
  return err
})
```

You could also write this as a test, i.e. have genValue return N% error, run it M times and ensure no errors bubble up, but with asserts, not a script.